### PR TITLE
[flutter_plugin_tools] Overhaul drive-examples

### PR DESCRIPTION
**Pull request by stuartmorgan**
Significantly restructures drive-examples:
- Migrates it to the new package-looping base command
- Enforces that only one platform is passed, since in practice multiple platforms never actually worked. (The logic is structured so that it will be easy to enable multi-platform if `flutter drive` gains multi-platform support.)
- Fixes the issue where `--ios` and `--android` were semi-broken, by doing explicit device targeting for them rather than relying on the default device being the right kind
- Extracts much of the logic to helpers so it's easier to understand the flow
- Removes support for a legacy integration test file structure that is no longer used
- Adds more test coverage; previously no failure cases were actually tested.

Fixes https://github.com/flutter/flutter/issues/85147
Part of https://github.com/flutter/flutter/issues/83413

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: script/tool/lib/src/drive_examples_command.dart [hunk 1]
```diff
@@ -43,213 +48,252 @@ class DriveExamplesCommand extends PluginCommand {
...
   @override
   final String description = 'Runs driver tests for plugin example apps.\n\n'
-      'For each *_test.dart in test_driver/ it drives an application with a '
-      'corresponding name in the test/ or test_driver/ directories.\n\n'
-      'For example, test_driver/app_test.dart would match test/app.dart.\n\n'
-      'This command requires "flutter" to be in your path.\n\n'
-      'If a file with a corresponding name cannot be found, this driver file'
-      'will be used to drive the tests that match '
-      'integration_test/*_test.dart.';
+      'For each *_test.dart in test_driver/ it drives an application with the '
```
Comment (gaaclarke) [inline script/tool/lib/src/drive_examples_command.dart:51]: s/the/

File: script/tool/lib/src/drive_examples_command.dart [hunk 2]
```diff
@@ -43,213 +48,252 @@ class DriveExamplesCommand extends PluginCommand {
...
+    }
+
+    String? androidDevice;
+    if (getBoolArg(kPlatformAndroid)) {
+      final List<String> devices = await _getDevicesForPlatform('android');
+      if (devices.isEmpty) {
+        printError('No Android devices available');
+        throw ToolExit(_exitNoAvailableDevice);
+      }
+      androidDevice = devices.first;
```
Comment (gaaclarke) [inline script/tool/lib/src/drive_examples_command.dart:89]: Should we warn if there are more than one device?

File: script/tool/lib/src/drive_examples_command.dart [hunk 3]
```diff
@@ -43,213 +48,252 @@ class DriveExamplesCommand extends PluginCommand {
...
     }
-    print('\n\n');
+    if (!testsRan) {
+      printError('No driver tests were run ($examplesFound example(s) found).');
+      errors.add('No tests ran (use --exclude if this is intentional).');
+    }
+    return errors;
+  }
+
+  Future<List<String>> _getDevicesForPlatform(String platform) async {
```
Comment (gaaclarke) [inline script/tool/lib/src/drive_examples_command.dart:201]: nit: Splitting this into parsing the results of `devices` and calling `devices` would make it more testable.  (Although it looks like you can use MockProcess to test it fine.)

File: script/tool/lib/src/drive_examples_command.dart [hunk 4]
```diff
@@ -43,213 +48,252 @@ class DriveExamplesCommand extends PluginCommand {
...
+        }
+      }
     }
-    if (isWindows) {
-      return isWindowsPlugin(plugin);
+    return tests;
+  }
+
+  /// Runs each file from [targets] using `flutter drive`, returning a list of
+  /// any failing test targets.
```
Comment (gaaclarke) [inline script/tool/lib/src/drive_examples_command.dart:267]: nit: if you are going to have a docstring, it helps to mention all the arguments.  It's kind of hard to guess what they all mean.

File: script/tool/test/drive_examples_command_test.dart [hunk 5]
```diff
@@ -35,52 +39,55 @@ void main() {
...
-      ]);
+    test('fails for iOS if no iOS devices are present', () async {
+      setMockFlutterDevicesOutput(hasIosDevice: false);
+      expect(
+        () => runCapturingPrint(runner, <String>['drive-examples']),
+        throwsA(isA<ToolExit>()),
+      );
+    });
 
+    test('fails if no platforms are provided', () async {
```
Comment (gaaclarke) [inline script/tool/test/drive_examples_command_test.dart:76]: Shouldn't this test have a device, but fail since there is no platform provided?

**Review by gaaclarke — APPROVED** [review]
The diff was a bit hard to parse, but it looks like all the old tests are still intact, just refactored.

File: script/tool/lib/src/drive_examples_command.dart (hunk 4 above)
Comment (stuartmorgan) [reply script/tool/lib/src/drive_examples_command.dart]: Done.

File: script/tool/test/drive_examples_command_test.dart (hunk 5 above)
Comment (stuartmorgan) [reply script/tool/test/drive_examples_command_test.dart]: This test was copypasta'd (it was supposed to be the Android version of the test above but I never actually changed it); fixed. I also did a sweep and fixed all tests that asserted `ToolExit`s to actually check the output as well so that they aren't failing for the wrong reasons (which was true of several more tests).

File: script/tool/lib/src/drive_examples_command.dart (hunk 1 above)
Comment (stuartmorgan) [reply script/tool/lib/src/drive_examples_command.dart]: Done

File: script/tool/lib/src/drive_examples_command.dart (hunk 2 above)
Comment (stuartmorgan) [reply script/tool/lib/src/drive_examples_command.dart:89]: I thought about adding more control over devices via flags, but I figured that absent anyone actually wanting that functionality I'd just do the simplest thing for now, and we can expand later (e.g., adding device choice flags to the command itself) if needed.

File: script/tool/lib/src/drive_examples_command.dart (hunk 3 above)
Comment (stuartmorgan) [reply script/tool/lib/src/drive_examples_command.dart:201]: For now relying on the indirect testing is fine; if we end up needing this in more places and pull this out as a utility method we can split it apart and do direct unit testing.

**Comment by stuartmorgan** [comment]
Landing on "red" since `submit-queue` isn't updating correctly.
